source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -117,34 +157,39 @@ void Filter::onGrpcClose() {
   stats_.streams_closed_.inc();
   // Successful close. We can ignore the stream for the rest of our request
   // and response processing.
-  // Use a switch here now because there will be more than two
-  // cases very soon.
-  switch (request_state_) {
-  case FilterState::HEADERS:
+  cleanupState();
+}
+
+void Filter::cleanupState() {
+  if (request_state_ != FilterState::IDLE) {
     request_state_ = FilterState::IDLE;
     decoder_callbacks_->continueDecoding();
-    break;
-  default:
-    // Nothing to do otherwise
-    break;
+  }
+  if (response_state_ != FilterState::IDLE) {
+    response_state_ = FilterState::IDLE;
+    encoder_callbacks_->continueEncoding();
   }
 }
 
-void Filter::sendImmediateResponse(const ImmediateResponse& response) {
+void Filter::sendImmediateResponse(const ImmediateResponse& response, bool on_decoding) {
   const auto status_code = response.has_status() ? response.status().code() : 200;
   const auto grpc_status =
       response.has_grpc_status()
           ? absl::optional<Grpc::Status::GrpcStatus>(response.grpc_status().status())
           : absl::nullopt;
+  const auto mutate_headers = [&response](Http::ResponseHeaderMap& headers) {
+    if (response.has_headers()) {
+      MutationUtils::applyHeaderMutations(response.headers(), headers);
+    }
+  };
 
-  decoder_callbacks_->sendLocalReply(
-      static_cast<Http::Code>(status_code), response.body(),
-      [&response](Http::ResponseHeaderMap& headers) {
-        if (response.has_headers()) {
-          MutationUtils::applyHeaderMutations(response.headers(), headers);
-        }
-      },
-      grpc_status, response.details());
+  if (on_decoding) {
+    decoder_callbacks_->sendLocalReply(static_cast<Http::Code>(status_code), response.body(),
```

Comment:
Does it matter if you use decoder/encode callbacks here? I don't recall TBH, but this seems strange that it would make a difference.

Reply:
It does seem strange but the API docs don't really say so I was trying to play it safe. Can we be sure that there will not be a difference in some future version of Envoy?

Reply:
These are the same inside of Envoy, there is no functional difference. The reason you have both is you have an encoder-decoder filter, whereas if you just had decoder (for example), you would have one. I think you can safely just use the decoder callbacks.

Reply:
OK -- I'll change that and resolve.